crates/driver/src/domain/competition/mod.rs

```diff
@@ -7,24 +23,81 @@ pub use {
     auction::Auction,
     order::Order,
     quote::Quote,
-    solution::{solve, Score, Solution, SolverTimeout},
+    solution::{Score, Solution, SolverTimeout},
 };
 
-use crate::domain::eth;
+/// An ongoing competition. There is one competition going on per solver at any
+/// time. The competition stores settlements to solutions generated by the
+/// driver, and allows them to be executed as requested later. The settlements
+/// expire after a certain amount of time, at which point trying to use them
+/// will return an `[Error::SolutionNotFound]`.
+#[derive(Debug)]
+pub struct Competition {
+    pub solver: Solver,
+    pub eth: Ethereum,
+    pub simulator: Simulator,
+    pub now: time::Now,
+    pub mempools: Vec<Mempool>,
+    pub settlements: Arc<DashMap<solution::Id, solution::Settlement>>,
+}
 
-/// The price of a token in wei. This represents how much wei is needed to buy
-/// 10**18 of another token.
-#[derive(Debug, Clone, Copy)]
-pub struct Price(pub eth::Ether);
+impl Competition {
+    /// Solve an auction as part of this competition.
+    pub async fn solve(&self, auction: &Auction) -> Result<(solution::Id, solution::Score), Error> {
+        let solution = self
+            .solver
+            .solve(
+                auction,
+                SolverTimeout::for_solving(auction.deadline, self.now)?,
+            )
+            .await?;
+        // TODO(#1009) Keep in mind that the driver needs to make sure that the solution
+        // doesn't fail simulation. Currently this is the case, but this needs to stay
+        // the same as this code changes.
+        let gas = solution
+            .simulate(&self.eth, &self.simulator, auction)
+            .await?;
+        let settlement = solution::Settlement::encode(&self.eth, auction, &solution).await?;
+        let score = settlement.score(&self.eth, auction, gas).await?;
+        let solution_id = solution::Id::random();
+        self.settlements.insert(solution_id, settlement);
+        // Remove the settlement after it's expired.
+        let settlements = Arc::clone(&self.settlements);
+        tokio::spawn(async move {
+            tokio::time::sleep(Self::expiration_time()).await;
+            settlements.remove(&solution_id);
+        });
```

Comment:
There is only one competition at a time. This means that from the perspective of the driver it gets a call to /solve and optionally a call to /execute afterwards. Then it gets a call to /solve again and so on. Based on this, we don't need to get rid of after a fixed time. We can keep it until the next call to /solve. This means we don't need a hashmap and can just store the most recently returned solution.

At least that's how I think this process works. Might have to clarify it in the openapi spec.

Reply:
It is not necessary because we can replace the stored solution on the next /solve call. There is no need to also replace after timeout.

Reply:
I think then you limit your driver to being run per deployment. One advantage of the `DashMap` is that you can, in theory, as an external solver use the same `driver` + `solver` instance to both production and staging. Since the implementation is quite simple, I think I'm fine with the `DashMap` (although I probably would have used a `Mutex<HashMap>` to avoid the external dependency since performance is not a concern here - the locking operation happens 0.2 times per second).

Reply:
> I think then you limit your driver to being run per deployment. 

That's true. If we want to support running for multiple deployments then this doesn't work. 

> DashMap vs Mutex<HashMap>

I also prefer Mutex<HashMap> when performance doesn't matter like in this case. No dependency and much simpler component.

Reply:
> But aren't production and staging always running on different networks?

No, we run the Mainnet/Production and Mainnet/Staging for example. A small solver team may want to run a single driver instance to reduce on node RPC costs for example.

> So I think @vkgnosis might be correct here

👍 - I think @vkgnosis's suggestion is good and it simplifies the code a bit. We can always revisit if we get a request for using a single solver + driver instance for multiple deployments.